### DISCUSSION TITLE
Add an equivalent to VB.NET's Handles clause

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#3085] [Language Ideas]
In VB I can say this:

```
Private Sub Page_Init(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Init
```

However in C# I can't add an event handler in this fashion; I have to add it in a separate line of code, which leads to fragmented code because the declaration of the event handler is separate from the registration of it as an event handler for a particular event:

```
private void Page_Init(object sender, EventArgs e)
...
base.Init += Page_Init;
```

I would like there to be added therefore an equivalent to VB's `Handles` clause, perhaps something like this:

```
private void Page_Init(object sender, EventArgs e) handles base.Init
```

**Comment by alrz** [comment]
https://github.com/dotnet/csharplang/issues/546

**Comment by DavidArno** [comment]
@alrz, that's an unrelated issue, I think.

**Comment by alrz** [comment]
sorry, I missed that this is about event handlers.

**Comment by 06needhamt** [comment]
I like this idea and would be willing to draw up a prototype if anyone is interested.

**Comment by HaloFour** [comment]
I'd suggest that this is intentional in C#, especially since it has been this way since both C# 1.0 and VB.NET 1.0.  The `Handles` clause has to introduce the same code, which means that the developer needs to know when the registration actually happens and when it's possible for the event to be raised before the handler has been subscribed.  Writing that code removes that question.  Event handlers feel like field initializers, which because they are executed prior to the call to the base constructor cannot refer to instance members.  But event handlers must reference instance members, both the event itself and the instance/method to which the event is attached.  VB.NET has always done field initialization post base constructor allowing for referencing instance members, like events and event handlers.  It's one of many philosophical differences between the two languages.  I also don't find it particularly onerous.  Designers and IDEs are already doing the majority of the work, particularly in those situations where you're likely to run into a lot of events, like UI frameworks.
